Firestore/core/src/model/value_util.cc

```diff
@@ -455,12 +457,61 @@ std::string CanonicalId(const google_firestore_v1_Value& value) {
   }
 }
 
-google_firestore_v1_Value DeepClone(google_firestore_v1_Value source) {
+std::string CanonicalId(const google_firestore_v1_ArrayValue& value) {
+  return CanonifyArray(value);
+}
+
+bool Contains(google_firestore_v1_ArrayValue haystack,
+              google_firestore_v1_Value needle) {
+  for (pb_size_t i = 0; i < haystack.values_count; ++i) {
+    if (Equals(haystack.values[i], needle)) {
+      return true;
+    }
+  }
+  return false;
+}
+
+google_firestore_v1_Value NullValue() {
+  google_firestore_v1_Value null_value{};
+  null_value.which_value_type = google_firestore_v1_Value_null_value_tag;
+  return null_value;
+}
+
+bool IsNullValue(const google_firestore_v1_Value& value) {
+  return value.which_value_type == google_firestore_v1_Value_null_value_tag;
+}
+
+google_firestore_v1_Value NaNValue() {
+  google_firestore_v1_Value nan_value{};
+  nan_value.which_value_type = google_firestore_v1_Value_double_value_tag;
+  nan_value.double_value = std::numeric_limits<double>::quiet_NaN();
+  return nan_value;
+}
+
+bool IsNaNValue(const google_firestore_v1_Value& value) {
+  return value.which_value_type == google_firestore_v1_Value_double_value_tag &&
+         isnan(value.double_value);
```

Comment:
Should be `std::isnan` from `cmath` instead.

`isnan` is also available because `cmath` also include `math.h`, but should not be directly used. Just learnt this today from Costa's code review :-)

Reply:
Very nice. Hopefully I can use this information in my next code review.